engine/src/test/java/io/zeebe/engine/processing/bpmn/subprocess/InterruptingEventSubprocessConcurrencyTest.java

```diff
@@ -0,0 +1,166 @@
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.zeebe.engine.processing.bpmn.subprocess;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.tuple;
+
+import io.zeebe.engine.util.EngineRule;
+import io.zeebe.engine.util.RecordToWrite;
+import io.zeebe.model.bpmn.Bpmn;
+import io.zeebe.model.bpmn.BpmnModelInstance;
+import io.zeebe.model.bpmn.builder.ProcessBuilder;
+import io.zeebe.protocol.impl.record.value.message.MessageRecord;
+import io.zeebe.protocol.record.intent.MessageIntent;
+import io.zeebe.protocol.record.intent.MessageSubscriptionIntent;
+import io.zeebe.protocol.record.intent.ProcessInstanceIntent;
+import io.zeebe.protocol.record.value.BpmnElementType;
+import io.zeebe.test.util.record.RecordingExporter;
+import java.util.Map;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class InterruptingEventSubprocessConcurrencyTest {
+
+  private static final String PROCESS_ID = "proc";
+  private static final String MSG_NAME = "messageName";
+
+  @Rule public final EngineRule engineRule = EngineRule.singlePartition();
+
+  @Test
+  // https://github.com/camunda-cloud/zeebe/issues/6552
+  public void shouldEndProcess() {
+    // given
+    final ProcessBuilder process = Bpmn.createExecutableProcess(PROCESS_ID);
+
+    process
+        .eventSubProcess("event_sub_proc")
+        .startEvent("event_sub_start")
+        .interrupting(true)
+        .message(b -> b.name(MSG_NAME).zeebeCorrelationKeyExpression("key"))
+        .endEvent("event_sub_end");
+
+    final BpmnModelInstance model =
+        process
+            .startEvent("start_proc")
+            .intermediateCatchEvent("catch")
+            .message(m -> m.name("msg").zeebeCorrelationKeyExpression("key"))
+            .exclusiveGateway()
+            .endEvent("end_proc")
+            .done();
+
+    engineRule.deployment().withXmlResource(model).deploy();
+
+    final long processInstanceKey =
+        engineRule
+            .processInstance()
+            .ofBpmnProcessId(PROCESS_ID)
+            .withVariables(Map.of("key", 123))
+            .create();
+
+    // when
+    RecordingExporter.messageSubscriptionRecords(MessageSubscriptionIntent.CREATED)
+        .withProcessInstanceKey(processInstanceKey)
+        .withMessageName(MSG_NAME)
+        .await();
+    RecordingExporter.messageSubscriptionRecords(MessageSubscriptionIntent.CREATED)
+        .withProcessInstanceKey(processInstanceKey)
+        .withMessageName("msg")
+        .await();
+
+    engineRule.writeRecords(
+        RecordToWrite.command()
+            .message(
+                MessageIntent.PUBLISH,
+                new MessageRecord().setName("msg").setCorrelationKey("123").setTimeToLive(0)),
+        RecordToWrite.command()
+            .message(
+                MessageIntent.PUBLISH,
+                new MessageRecord().setName(MSG_NAME).setCorrelationKey("123").setTimeToLive(0)));
```

Comment:
Please propose a better test or approach, otherwise I see no way in improving here. What I can think of is to pause the processing in a certain point and write the commands on the stream, but this is something you didn't wanted. I wouldn't go the way of writing the complete (all) records to the stream.

Reply:
I think if we add the generated event sub process to the property based tests then we have also another good proof that it is working, because before most of the test have failed.

Reply:
I have no intermediate solution. We could write more records explicitly. But this is more complicated. We can write such a test later. Let's create an issue to increase the coverage.

The property tests should be good enough for now :+1: